libraries/chain/include/eosio/chain/webassembly/interface.hpp

```diff
@@ -0,0 +1,581 @@
+         void eosio_assert_message(bool, legacy_array_ptr<const char>);
+         void eosio_assert_code(bool, uint64_t);
+         void eosio_exit(int32_t);
+
+         REGISTER_CF_HOST_FUNCTION(abort)
+         REGISTER_LEGACY_CF_HOST_FUNCTION(eosio_assert)
+         REGISTER_LEGACY_CF_HOST_FUNCTION(eosio_assert_message)
+         REGISTER_CF_HOST_FUNCTION(eosio_assert_code)
+         REGISTER_CF_HOST_FUNCTION(eosio_exit)
+
+         // action api
+         int32_t read_action_data(legacy_array_ptr<char>) const;
+         int32_t action_data_size() const;
+         name current_receiver() const;
+         void set_action_return_value(legacy_array_ptr<char>);
+
+         REGISTER_LEGACY_CF_HOST_FUNCTION(read_action_data);
+         REGISTER_CF_HOST_FUNCTION(action_data_size);
+         REGISTER_CF_HOST_FUNCTION(current_receiver);
+         REGISTER_LEGACY_HOST_FUNCTION(set_action_return_value);
+
+         // console api
+         void prints(null_terminated_ptr);
+         void prints_l(legacy_array_ptr<const char>);
+         void printi(int64_t);
+         void printui(uint64_t);
+         void printi128(const __int128&);
+         void printui128(const unsigned __int128&);
+         void printsf(float);
+         void printdf(double);
+         void printqf(const float128_t&);
+         void printn(name);
+         void printhex(legacy_array_ptr<const char>);
+
+         REGISTER_LEGACY_CF_HOST_FUNCTION(prints);
+         REGISTER_LEGACY_CF_HOST_FUNCTION(prints_l);
+         REGISTER_LEGACY_CF_HOST_FUNCTION(printi);
+         REGISTER_LEGACY_CF_HOST_FUNCTION(printui);
+         REGISTER_CF_HOST_FUNCTION(printi128);
+         REGISTER_CF_HOST_FUNCTION(printui128);
+         REGISTER_LEGACY_CF_HOST_FUNCTION(printsf);
+         REGISTER_LEGACY_CF_HOST_FUNCTION(printdf);
+         REGISTER_CF_HOST_FUNCTION(printqf);
+         REGISTER_LEGACY_CF_HOST_FUNCTION(printn);
+         REGISTER_LEGACY_CF_HOST_FUNCTION(printhex);
+
+         // database api
+         // primary index api
+         int32_t db_store_i64(uint64_t, uint64_t, uint64_t, uint64_t, legacy_array_ptr<const char>);
+         void db_update_i64(int32_t, uint64_t, legacy_array_ptr<const char>);
+         void db_remove_i64(int32_t);
+         int32_t db_get_i64(int32_t, legacy_array_ptr<char>);
+         int32_t db_next_i64(int32_t, uint64_t&);
+         int32_t db_previous_i64(int32_t, legacy_ptr<uint64_t>);
+         int32_t db_find_i64(uint64_t, uint64_t, uint64_t, uint64_t);
+         int32_t db_lowerbound_i64(uint64_t, uint64_t, uint64_t, uint64_t);
+         int32_t db_upperbound_i64(uint64_t, uint64_t, uint64_t, uint64_t);
+         int32_t db_end_i64(uint64_t, uint64_t, uint64_t);
+
+         REGISTER_LEGACY_HOST_FUNCTION(db_store_i64);
+         REGISTER_LEGACY_HOST_FUNCTION(db_update_i64);
+         REGISTER_LEGACY_HOST_FUNCTION(db_remove_i64);
+         REGISTER_LEGACY_HOST_FUNCTION(db_get_i64);
+         REGISTER_HOST_FUNCTION(db_next_i64);
```

Comment:
I'm concerned that references like `uint64_t&` may not be treated the same under the new system as they were under the old system. There may be a need for a `legacy_ref` wrapper type.

Reply:
No need for `legacy_ref`. Just need to be consistent about using `legacy_ptr<uint64_t>` like in `db_previous_i64` rather than using `uint64_t&`.